### ISSUE TITLE
[Bug] [Table] Component wrapped by React.memo passed into `components` of Table will get unmount every time Table renders.

### ISSUE THREAD

**Issue by NE-SmallTown** [ant-design/ant-design#32897]
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/antd-table-mergeobject-bug-j9dse)

### Steps to reproduce
At [ `rc-table`](https://github.com/react-component/table/blob/dd480b2d4a56432d09b942872366cb7190d6cafc/src/Table.tsx#L226) , it will recreate the `components` prop passed into the `Table`, it will use the `mergeObject` helper function. This will make our memo or forwardRef Component in `components`(like `components.body.wrapper`) **become a new object**, so the Component will be unmounted and recreate & remount. The reason is below:

```jsx
function mergeObject(...objects) {
  const merged = {};

  function fillProps(obj, clone) {
    if (clone) {
      Object.keys(clone).forEach(key => {
        const value = clone[key];
        if (value && typeof value === 'object') {
          obj[key] = obj[key] || {};
          fillProps(obj[key], value);
        } else {
          obj[key] = value;
        }
      });
    }
  }

  objects.forEach(clone => {
    fillProps(merged, clone);
  });

  return merged;
}

// try this:

function FooComponent() {}

const BarComponent = React.memo(function BarComponent() {})

const BazComponent = React.forwardRef(function BazComponent() {})

const components = {
    body: {
        wrapper: FooComponent,
        row: BarComponent,
      	cell: BazComponent
    }
}

const mergedComponents = mergeObject(components, {});

console.log('mergedComponents.body.wrapper === FooComponent', mergedComponents.body.wrapper === FooComponent);
console.log('mergedComponents.body.row === BarComponent', mergedComponents.body.row === BarComponent);
console.log('mergedComponents.body.cell === BazComponent', mergedComponents.body.cell === BazComponent);
```

![image](https://user-images.githubusercontent.com/18418010/141996677-5df6e59d-83c3-413f-8534-f454b9927ec2.png)

So, if we have codes that will `setState` in the component at some situations, it will throw Error: `Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the undefined component.` 

### What is expected?
Component won't be unmounted.

### What is actually happening?
Component be unmounted.

| Environment | Info |
|---|---|
| antd | 4.17.0 |
| React | latest |
| System | MacOS |
| Browser | Chrome 95 |

---
### Possible solution

I think we can just remove the `mergedComponents` logic and use `components` directly instead. But if you want keep that, you probably need do more precise checks like [ReactIs.isMemo](https://github.com/facebook/react/blob/main/packages/react-is/src/ReactIs.js#L131)  or [ReactIs.isForwardRef](https://github.com/facebook/react/blob/4ff5f5719b348d9d8db14aaa49a48532defb4ab7/packages/react-is/src/ReactIs.js#L122) in the `mergeObject` function.

**Comment by afc163** [comment]
Is https://github.com/ant-design/ant-design/issues/32670 and https://github.com/ant-design/ant-design/issues/23763 helpful?

**Comment by NE-SmallTown** [comment]
No, I have lifted the Component to the top, I think this is not about that issue. You can look through my code above

**Comment by NE-SmallTown** [comment]
No, I have lifted the Component to the top, I think this is not about that issue. You can look through my code above

**Comment by linxianxi** [comment]
+1

**Comment by EmilyyyLiu** [comment]
我创建了一个 Demo 来测试这个问题，但无法复现：

**Demo**: https://codesandbox.io/p/devbox/ji-ben-yong-fa-antd-6-4-3-forked-hhqn6g?workspaceId=ws_T98hW5dcqudRAGy3wbJh83

点击按钮触发重渲染后，wrapper 组件没有发生卸载重建，问题应该已在后续版本中修复。

这个问题和 #32922 是同一个问题，已在最新版本中修复。

如果还有问题，欢迎继续讨论。如果没有问题的话，这个 issue 是否可以关闭？ ✅
